### ISSUE TITLE
Improve testability

### ISSUE THREAD

**Issue by beorn7** [prometheus/client_golang#58]
First of all, there is the public Registry interface tracked in https://github.com/prometheus/client_golang/issues/46

But there is more:

Two things come to my mind here, from the email thread:

(1)
Another utility function could compare two collectors (or a collector
to a list of template metrics) and verify that they return the same
set of metrics. In that way, you could simply create the expected
metrics and verify that the collector to be tested collects the same
metrics.
[...]
But one of my suggestions was to have a comparison utility function
that would abstract away the order (as the Prometheus registry does
it, too).
Something like

```
testutil.Collects(myCollector, []prometheus.Metric{
    prometheus.MustNewConstMetric(...),
}
```

(2)

``` go
func readCounter(m prometheus.Counter) float64 {
    pb := &dto.Metric{}
    m.Write(pb)
    return pb.GetCounter().GetValue()
}
```

Could be done in a generic way, like this:

``` go
package testutil

func ToFloat64(m prometheus.Metric) float64 {
    pb := &dto.Metric{}
    m.Write(pb)
    if pb.Gauge != nil {
        return pb.Gauge.GetValue()
    }
    if pb.Counter != nil {
        return pb.Counter.GetValue()
    }
    if pb.Untyped != nil {
        return pb.Untyped.GetValue()
    }
    return math.NaN()
}
```

**Comment by brian-brazil** [comment]
https://github.com/prometheus/client_java/blob/master/simpleclient/src/main/java/io/prometheus/client/CollectorRegistry.java#L102 is one way that the java_client makes things easier to test

**Comment by grobie** [comment]
:+1:

**Comment by beorn7** [comment]
This represents the first part of test tools, those that can be implemented without the more invasive changes in 0.10. For the rest, see #230

**Comment by beorn7** [comment]
Plan is to adopt the [kube-state-metrics testutils package](https://github.com/kubernetes/kube-state-metrics/blob/2e30678a4d4d055ce13a095c2ec8d14eef353570/pkg/collectors/testutils/testutils.go). @brancz will create a PR. See #422 for details.
